### ISSUE TITLE
db: `RangeKeyChanged()` does not fire when iterator is exhausted

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/pebble#2199]
Related to https://github.com/cockroachdb/cockroach/issues/94041.

CRDB test case which fails on `RangeKeyChanged()`:

```go
func TestRangeKeyChanged(t *testing.T) {
	eng := createTestPebbleEngine()
	defer eng.Close()

	require.NoError(t, eng.PutMVCCRangeKey(
		MVCCRangeKey{StartKey: roachpb.Key("a"), EndKey: roachpb.Key("b"), Timestamp: hlc.MinTimestamp},
		MVCCValue{}))

	db := eng.(*Pebble).db
	iter := db.NewIter(&pebble.IterOptions{
		KeyTypes: pebble.IterKeyTypeRangesOnly,
	})
	defer iter.Close()
	require.True(t, iter.SeekGE([]byte{}))
	require.True(t, iter.RangeKeyChanged())
	require.False(t, iter.Next())
	require.True(t, iter.RangeKeyChanged())
}
```

**Comment by jbowens** [comment]
Do we rely on this behavior? Almost all `pebble.Iterator` methods' behavior is undefined if `!Valid()`

**Comment by erikgrinaker** [comment]
I believe we specified it as such when we originally added it. It's documented here:

https://github.com/cockroachdb/cockroach/blob/a269b145be692473716ecaffb6a927d68fbd30fe/pkg/storage/engine.go#L206-L208

It comes in handy occasionally. Over in https://github.com/cockroachdb/cockroach/pull/93753 we call it without calling Valid to avoid redundant Valid calls in a hot path. It also would've been handy below, where we need to know when we stepped _off_ of a range key (including stepping to the end of the iterator):

https://github.com/cockroachdb/cockroach/blob/a434560ea69959aae584a32f8044d1840d821d70/pkg/storage/sst.go#L829-L833

Furthermore, if we don't return `true` when we exhaust the iterator, it isn't clear what it'd return when we stepped back onto the range key -- does it return true or false? This can be problematic in code that's given an already-used iterator, if a seek operation on an invalid iterator lands on the range key that the iterator was positioned on before being exhausted, in which case the caller might expect `RangeKeyChanged` to always return `true`. The `pebbleMVCCScanner` in 22.1 relies on this assumption to enable point synthesis. So I think we want to be clear about the semantics here.

**Comment by jbowens** [comment]
> I believe we specified it as such when we originally added it. It's documented here:

I believe we defined the return value if the current position is `Valid()` but the previous position was exhausted. I don't believe we ever defined it for `!Valid()`. 

> Furthermore, if we don't return true when we exhaust the iterator, it isn't clear what it'd return when we stepped back onto the range key -- does it return true or false? 

I think this is exactly what we did previously define.

We can make the change to make RangeKeyChanged return false if `!Valid()`, it'll just be a bit of a departure from the rest of the `pebble.Iterator` interface which expects the caller to not invoke any methods related to positional state if the iterator is invalid.
